File: C/src/Week-1-2_Data-Structures/Stack/Using LinkedList/Basic/Operations/main.c
#include <stdio.h>
#include <stdlib.h>

struct node
{
    int data;
    struct node *link;
};


struct node *top = 0;

void push(int);
void display();
void peek();
void pop();

int main(void)
{
    push(3);
    push(4);
    push(5);
    pop();
    display();
    peek();

    return (0);
}

void push(int x)
{
   struct node *newnode;
    newnode = malloc(sizeof(struct node));
    newnode -> data = x;
    newnode -> link = top;
    top = newnode;
}

void display()
{
    struct node *temp = top;
    if ( top == 0 )
        printf("Empty stack Linked List");
    else
    {
        while (temp != 0)
        {
            printf("%d\n", temp->data);
            temp = temp->link;
        }
    }
}

void peek()
{
    if ( top == 0 )
        printf("Empty stack Linked List");
    else
        printf("Top: %d\n", top->data);
}

void pop()
{
    struct node *temp = top;
    if ( top == 0 )
        printf("Empty stack Linked List");
    else
    {
        printf("value: %d\n", top->data);
        top = top->link;
        free(temp);
    }  
}

File: Python/src/Week-1-2_Data-Structures/Stack/Using LinkedList/Basic/Operations/main.py
class Node:
    def __init__(self, value: int) -> None:
        self.value = value
        self.next = None


class Stack:
    def __init__(self, value: int) -> None:
        new_node = Node(value)
        self.top = new_node
        self.height = 1

    def print_stack(self) -> None:
        temp = self.top
        while temp is not None:
            print(temp.value, end="\t")
            temp = temp.next
        print()

    def push(self, value: int) -> bool:
        new_node = Node(value)
        if self.height == 0:
            self.top = new_node
        else:
            new_node.next = self.top
            self.top = new_node
        self.height += 1

    def pop(self) -> Node:
        if self.top is None:
            return None
        temp = self.top
        self.top = self.top.next
        temp.next = None
        self.height -= 1
        return temp


my_stack = Stack(4)
my_stack.push(3)
my_stack.print_stack()
my_stack.pop()
my_stack.print_stack()
